crates/ruff_linter/src/rules/pycodestyle/rules/blank_lines.rs

```diff
@@ -0,0 +1,732 @@
+                )));
+
+                diagnostics.push(diagnostic);
+            }
+
+            if line.preceding_blank_lines < BLANK_LINES_TOP_LEVEL
+                && is_top_level_token(&self.previous_unindented_token)
+                && line.indent_level == 0
+                && !line.is_comment_only
+                && !is_top_level_token_or_decorator(&line.first_token)
+            {
+                // E305
+                let mut diagnostic = Diagnostic::new(
+                    BlankLinesAfterFunctionOrClass(line.blank_lines),
+                    line.first_token_range,
+                );
+
+                diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
+                    stylist
+                        .line_ending()
+                        .as_str()
+                        .to_string()
+                        .repeat((BLANK_LINES_TOP_LEVEL - line.blank_lines) as usize),
+                    locator.line_start(line.first_token_range.start()),
+                )));
+
+                diagnostics.push(diagnostic);
+            }
+
+            if line.preceding_blank_lines == 0
+            // Only apply to nested functions.
+                && matches!(self.fn_status, Status::Inside(_))
+                && is_top_level_token_or_decorator(&line.first_token)
+                // Allow following a decorator (if there is an error it will be triggered on the first decorator).
+                && !matches!(self.follows, Follows::Decorator)
+                // The class's docstring can directly precede the first function.
+                && !matches!(self.follows, Follows::Docstring)
+                // Do not trigger when the def/class follows an "indenting token" (if/while/etc...).
+                && prev_indent_level.is_some_and(|prev_indent_level| prev_indent_level >= line.indent_level)
+                // Allow groups of one-liners.
+                && !(matches!(self.follows, Follows::Def) && !matches!(line.last_token, Tok::Colon))
+            {
+                // E306
+                let mut diagnostic = Diagnostic::new(
+                    BlankLinesBeforeNestedDefinition(line.blank_lines),
+                    line.first_token_range,
+                );
+
+                diagnostic.set_fix(Fix::safe_edit(Edit::insertion(
+                    stylist.line_ending().as_str().to_string(),
+                    locator.line_start(line.first_token_range.start()),
+                )));
+
+                diagnostics.push(diagnostic);
+            }
+        }
+
+        match line.first_token {
+            Tok::Class => {
+                if matches!(self.class_status, Status::Outside) {
+                    self.class_status = Status::Inside(line.indent_level + indent_size);
+                }
+                self.follows = Follows::Other;
+            }
+            Tok::At => {
+                self.follows = Follows::Decorator;
+            }
+            Tok::Def | Tok::Async => {
```

Comment:
As far as I understand it's impossible for `first_token` to be the `Async` token because we explicitly skip it in the `LinePreprocessor`. 

```suggestion
            Tok::Def => {
```

It could make sense to have a custom `FirstToken` enum that only has the variants that we care about in the `BlankLines` implementation (could it be the same as `Follows`?). It would move the detection of the supported kinds into the `LinePreprocessor` and make it easier to reason about if all relevant tokens have been handled in the `check_line` function

Reply:
I've removed the `Async` in 531ce0ed5. It's a left-over from before I realized that `with async` would cause trouble and started to ignore asyncs.

> It could make sense to have a custom FirstToken enum that only has the variants that we care about in the BlankLines implementation (could it be the same as Follows?). It would move the detection of the supported kinds into the LinePreprocessor and make it easier to reason about if all relevant tokens have been handled in the check_line function

I don't think it would make `check_line` easier to reason about since it checks the violations one by one instead of matching on the `TokenKind` (that's probably not optimal in terms of speed, but it makes it much, much easier to see what leads to a given violation being raised).